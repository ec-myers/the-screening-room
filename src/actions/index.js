export const setMovies = movies => ({
  type: 'SET_MOVIES',
  movies
})

export const setWallpapers = wallpapers => ({
  type: 'SET_WALLPAPERS',
  wallpapers
})

export const setLoading = bool => ({
  type: 'IS_LOADING',
  bool
})

export const hasError = error => ({
  type: 'HAS_ERROR',
  error
})

export const hasEmailError = error => ({
  type: 'HAS_EMAIL_ERROR',
  error
})

export const toggleFavorite = title => ({
  type: 'TOGGLE_FAVORITE',
  title
})

export const setFavorites = favorites => ({
  type: 'SET_FAVORITES',
  favorites
})

export const addFavorite = newFavorite => ({
  type: 'ADD_FAVORITE',
  newFavorite
})

export const setUser = user => ({
  type: 'USER',
  user
});

export const logoutUser = () => ({
  type: 'USER_LOGOUT'
});
