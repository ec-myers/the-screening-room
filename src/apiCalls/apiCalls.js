const baseUrl = 'https://api.themoviedb.org/3/movie/now_playing?api_key=';
const imageBaseUrl = 'https://image.tmdb.org/t/p/original';
const genreBaseUrl = 'https://api.themoviedb.org/3/genre/';
const apiKey = '149174d30ba0677b5219f8786eaaaaa7';

export const getMovies = async () => {
  const response = await fetch(`${baseUrl}${apiKey}`);
  const data = await response.json();
  const results = data.results;
  const cleanedMovies = await results.map(async result => {
    const {
      adult,
      backdrop_path,
      genre_ids,
      title,
      overview,
      poster_path,
      release_date,
      vote_average
    } = result;
    return {
      rRated: adult,
      wallpaper: `${imageBaseUrl}${backdrop_path}`,
      genre: await getGenresData(genre_ids),
      title,
      overview,
      poster: `${imageBaseUrl}${poster_path}`,
      release_date,
      rating: vote_average,
      favorite: false
    };
  });
  return await Promise.all(cleanedMovies);
};

const getGenresData = async genreIds => {
  const genres = await genreIds.map(id => {
    return getGenres(id);
  });
  return await Promise.all(genres);
};

const getGenres = async id => {
  const response = await fetch(`${genreBaseUrl}${id}?api_key=${apiKey}`);
  const data = await response.json();
  return await data.name;
};

export const createUser = async newUser => {
  const url = 'http://localhost:3001/api/v1/users';
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(newUser)
  };
  const response = await fetch(url, options);
  if (!response.ok) {
    throw new Error('Email address already in use');
  }
  return response.json();
};

export const selectUser = async recurrentUser => {
  const url = 'http://localhost:3001/api/v1/login';
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(recurrentUser)
  };
  const response = await fetch(url, options);
}
  if (!response.ok) {
    throw new Error('Email and password do not match');
  }
  return response.json();
};

export const getFavorites = async userId => {
  const url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites`;
  const response = await fetch(url);
  const favorites = await response.json();
  console.log(favorites)
  return favorites;
}

export const postFavorite = async (movie, userId) => {
  const url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites`;
  const options = {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(movie)
  }
  const response = await fetch(url, options)
  const newFavorite = await response.json();
  return newFavorite;
}