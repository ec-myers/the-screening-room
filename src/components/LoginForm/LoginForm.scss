.form-model {
  margin: auto;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;

  p {
    color: rgb(43, 43, 43);
    width: 70%;
    text-align: center;
    height: 37px;
  }

  .login-error,
  .email-error {
    color: red;
    width: 100%;
  }

  div {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    height: 60px;
    margin-bottom: 50px;
  }

  .Form__label {
    font-size: 15px;
    color: gray;
  }

  ::placeholder {
    color: rgb(226, 226, 226);
  }

  input {
    border: none;
    border-bottom: 1px solid rgb(170, 170, 170);
    font-size: 18px;
    width: 270px;
    padding: 3px;

    &:focus {
      outline: none;
    }
  }

  button {
    font-size: 16px;
    border: none;
    background-color: rgb(43, 43, 43);
    border-radius: 4px;
    padding: 10px 20px;
    cursor: pointer;
    color: white;
    transition: 0.15s;

    &:focus {
      outline: none;
    }

    &:hover {
      background-color: rgb(0, 0, 0);
      transition: 0.15s;
    }
  }

  .Form__input--error {
    border-bottom: 1px solid red;
  }
}

.login-message {
  margin-bottom: 60px;
}
