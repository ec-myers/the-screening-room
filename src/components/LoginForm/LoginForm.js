import React, { Component } from 'react';
import { selectUser } from '../../apiCalls/apiCalls';
import './LoginForm.scss';
import { Link } from 'react-router-dom';

class LoginForm extends Component {
  constructor() {
    super();
    this.state = {
      email: '',
      password: ''
    };
  }

  handleChange = event => {
    this.setState({
      [event.target.name]: event.target.value
    })
  }

  handleClick = async () => {
    await selectUser(this.state)
    this.clearInputs();
  }

  clearInputs = () => {
    const stateKeys = Object.keys(this.state)
    stateKeys.forEach( stateKey => this.setState({ 
      [stateKey]: '' 
    }))
  }

  handleSubmit = event => {
    event.preventDefault();
  }

  render() {
    const { email, password } = this.state;
    return (
      <form className='form-model' onSubmit={(e) => this.handleSubmit(e)}>
        <p className="login-message">Sign in with your email to keep track of your favorite movies.</p>
        <div>
        <label htmlFor='email' className='Form__label'>
          Your email
        </label>
        <input
          id='email'
          type='text'
          placeholder='your@email.com'
          className='Form__input'
          name='email'
          value={email}
          onChange={(e) => this.handleChange(e)}
        ></input>
        </div>
        <div>
        <label htmlFor='password' className='Form__label'>
          Your password
        </label>
        <input
          id='password'
          type='password'
          className='Form__input'
          name='password'
          placeholder='Must have at least 8 characters'
          value={password}
          onChange={(e) => this.handleChange(e)}
        ></input>
        </div>
        <Link to='/'>
        <button 
          type='submit' 
          className='Form__button'
          onClick={(e) => this.handleClick(e)}
        >
          Continue
        </button>
        </Link>
      </form>
    );
  }
}

// mapStateToProps = (state) => ({
//   email: state.email,
//   password: state.password
// });

export default LoginForm;
// export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);