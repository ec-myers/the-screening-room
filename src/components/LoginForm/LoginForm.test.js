import React from 'react';
import { shallow } from 'enzyme';
import { LoginForm, mapStateToProps, mapDispatchToProps } from './LoginForm';
import { setUser, hasError, setFavorites } from '../../actions/index';
import { selectUser, getFavorites } from '../../apiCalls/apiCalls';

jest.mock('../../apiCalls/apiCalls');

describe('LoginForm', () => {
  let wrapper;
  let errMsg = jest.fn();
  let mockEventName = {
    target: {
      name: 'email',
      value: 'elyse@noneofyourbusiness.com'
    }
  };
  let mockEventPassword = {
    target: {
      name: 'password',
      value: 'testing123'
    }
  };

  beforeEach(() => {
    wrapper = shallow(<LoginForm errMsg={errMsg} />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update form state when handleChange is called', () => {
    const expectedEmail = 'elyse@noneofyourbusiness.com';
    const expectedPassword = 'testing123';

    wrapper.instance().handleChange(mockEventName);
    wrapper.instance().handleChange(mockEventPassword);
    expect(wrapper.state('email')).toEqual(expectedEmail);
    expect(wrapper.state('password')).toEqual(expectedPassword);
  });

  it('should reset state when clearInputs is called', () => {
    const expected = { email: '', password: '', isLoggedIn: '' };

    wrapper
      .instance()
      .setState({
        email: 'elyse@noneofyourbusiness.com',
        password: 'testing123'
      });
    wrapper.instance().clearInputs();
    expect(wrapper.state()).toEqual(expected);
  });

  it('should call the selectUser fetch when handleClick is called', () => {
    wrapper.instance().handleClick();
    expect(selectUser).toHaveBeenCalled();
  });

  it.skip('should update the value of isLoggedIn when handleClick is called', () => {
    expect(wrapper.state('isLoggedIn')).toEqual(false);
    wrapper.instance().handleClick();
    expect(wrapper.state('isLoggedIn')).toEqual(true);
  });

  it('should call the getFavorites fetch when findUserFavorites is called', () => {
    let user = {
      id: 1,
      name: 'Alan',
      email: 'alan@turing.io'
    };
    wrapper.instance().findUserFavorites(user);
    expect(getFavorites).toHaveBeenCalled();
  });
});

describe('mapStateToProps', () => {
  it('should return an object with a movies array, an error message and isLoading property', () => {
    const mockState = {
      movies: [{ id: 1 }, { id: 2 }, { id: 3 }],
      errMsg: '',
      isLoading: false
    };

    const expected = {
      movies: [{ id: 1 }, { id: 2 }, { id: 3 }],
      errMsg: '',
      isLoading: false
    };

    const mappedProps = mapStateToProps(mockState);

    expect(mappedProps).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  let user = {
    name: 'Vanessa Randall',
    email: 'vanessa.randall@doane.edu',
    password: 'Password123',
    id: 2
  };

  let favorites = [
    {
      id: 63,
      movie_id: 420809,
      user_id: 1,
      title: 'Maleficent: Mistress of Evil',
      poster_path:
        'https://image.tmdb.org/t/p/original/tBuabjEqxzoUBHfbyNbd8ulgy5j.jpg',
      release_date: '2019-10-18',
      vote_average: '7.2',
      overview:
        'Maleficent and her goddaughter Aurora begin to question the complex family ties that bind them as they are pulled in different directions by impending nuptials, unexpected allies, and dark new forces at play.',
      favorite: true
    },
    {
      id: 63,
      movie_id: 420809,
      user_id: 1,
      title: 'Maleficent: Mistress of Evil',
      poster_path:
        'https://image.tmdb.org/t/p/original/tBuabjEqxzoUBHfbyNbd8ulgy5j.jpg',
      release_date: '2019-10-18',
      vote_average: '7.2',
      overview:
        'Maleficent and her goddaughter Aurora begin to question the complex family ties that bind them as they are pulled in different directions by impending nuptials, unexpected allies, and dark new forces at play.',
      favorite: true
    }
  ];

  const mockDispatch = jest.fn();

  it('should call dispatch with a setUser action when handleClick is called', () => {
    const actionToDispatch = setUser(user);

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setUser(user);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('should call dispatch with a setFavorites action when handleClick is called', () => {
    const actionToDispatch = setFavorites(favorites);

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setFavorites(favorites);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('should call dispatch with a hasError action when handleClick is called', () => {
    const message = '';
    const actionToDispatch = hasError(message);

    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.hasError(message);

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});
