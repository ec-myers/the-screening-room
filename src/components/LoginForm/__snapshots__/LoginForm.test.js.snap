// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match the snapshot 1`] = `
<form
  className="form-model"
  onSubmit={[Function]}
>
  <p
    className="login-message login-error"
  >
    string
  </p>
  <div>
    <label
      className="Form__label"
      htmlFor="email"
    >
      Your email
    </label>
    <input
      className="Form__input--error Form__input"
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="your@email.com"
      type="text"
      value=""
    />
  </div>
  <div>
    <label
      className="Form__label"
      htmlFor="password"
    >
      Your password
    </label>
    <input
      className="Form__input--error Form__input"
      id="password"
      name="password"
      onChange={[Function]}
      placeholder="Must have at least 8 characters"
      type="password"
      value=""
    />
  </div>
  <button
    className="Form__button"
    type="submit"
  >
    Continue
  </button>
</form>
`;
